src/pip/_internal/operations/install/wheel.py

```diff
@@ -734,9 +705,8 @@ def _generate_file(path: str, **kwargs: Any) -> Iterator[BinaryIO]:
     record_path = os.path.join(dest_info_dir, 'RECORD')
 
     with _generate_file(record_path, **csv_io_kwargs('w')) as record_file:
-        # The type mypy infers for record_file is different for Python 3
-        # (typing.IO[Any]) and Python 2 (typing.BinaryIO). We explicitly
-        # cast to typing.IO[str] as a workaround.
+        # Explicitly cast to typing.IO[str] as a workaround for the mypy error:
+        # "writer" has incompatible type "BinaryIO"; expected "_Writer"
```

Comment:
Hmm, wait, something is wrong. `_generate_file` always returns a bytes-based file object, so we shouldn’t cast it to `IO[str]`. Does `IO[bytes]` work for `csv.writer`?

Reply:
> _generate_file always returns a bytes-based file object

Not always. It depends on the `mode` argument passed through to `adjacent_tmp_file` and `tempfile.NamedTemporaryFile`. Here mode is `w` so the IO object handles text not bytes.

> Does IO[bytes] work for csv.writer?

No it does not. `csv.writer` requires its underlying file be opened in text mode.

---

I'm not sure how to modify the type signatures of `adjacent_tmp_file` and `_generate_file` in order to make mypy happy. I'd be very open to some help.

Reply:
> Not always. It depends on the `mode` argument […]

Sorry, I meant the *type hint* of `_generate_file` says it always returns a bytes-based file object, and we should fix that if it’s not the case.

Reply:
I think how file built-in `open` does it is to do an overload that checks `mode` on a bunch of `Literal`s. If it’s too cumbersome, maybe we should do a `cast` in `_generate_file` instead with a comment that says this is correct and doing things “properly” is to much work and not worth it.

Reply:
After a month, seems like there isn't interest in following the typing fix. I've put the WIP in a branch if someone wants to pick it up later: https://github.com/jdufresne/pip/tree/generate-file

Without more progress there, I think the original PR offers and improvement by dropping the last internal use of six.